File: src/main/java/com/linebot/config/RedisConfiguration.java
package com.linebot.config;

import com.linebot.model.UserStatus;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.serializer.GenericJackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.StringRedisSerializer;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.connection.RedisConnectionFactory;

@Configuration
public class RedisConfiguration {
    @Bean
    public RedisTemplate<String, UserStatus> redisTemplate(RedisConnectionFactory connectionFactory) {
        RedisTemplate<String, UserStatus> redisTemplate = new RedisTemplate<String, UserStatus>();
        redisTemplate.setConnectionFactory(connectionFactory);
        redisTemplate.setKeySerializer(new StringRedisSerializer());
        redisTemplate.setValueSerializer(new GenericJackson2JsonRedisSerializer("UserStatus"));
        redisTemplate.setHashKeySerializer(redisTemplate.getKeySerializer());
        redisTemplate.setHashValueSerializer(redisTemplate.getValueSerializer());
        return redisTemplate;
    }
}

File: src/main/kotlin/com/linebot/config/RedisConfiguration.kt
package com.linebot.config

import com.linebot.model.UserStatus
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.serializer.Jackson2JsonRedisSerializer
import org.springframework.data.redis.serializer.StringRedisSerializer

@Configuration
class RedisConfiguration {
    @Bean
    fun userStatusRedisTemplate(connectionFactoryArg: RedisConnectionFactory): RedisTemplate<String, UserStatus> {
        return buildRedisTemplate(connectionFactoryArg, UserStatus::class.java)
    }

    @Bean
    fun nonceRedisTemplate(connectionFactoryArg: RedisConnectionFactory): RedisTemplate<String, String> {
        return buildRedisTemplate(connectionFactoryArg, String::class.java)
    }

    private fun <T> buildRedisTemplate(
        connectionFactoryArg: RedisConnectionFactory,
        clazz: Class<T>
    ): RedisTemplate<String, T> {
        return RedisTemplate<String, T>()
            .apply {
                connectionFactory = connectionFactoryArg
                keySerializer = StringRedisSerializer()
                valueSerializer = Jackson2JsonRedisSerializer(clazz)
                hashKeySerializer = this.keySerializer
                hashValueSerializer = this.valueSerializer
            }
    }
}
